tests/importer/importer.cpp

```diff
@@ -0,0 +1,761 @@
+    auto model3 = libcellml::Model::create("model3");
+    auto model4 = libcellml::Model::create("model4");
+
+    EXPECT_TRUE(importer->addModel(model1, "a_model_the_first.cellml"));
+    EXPECT_TRUE(importer->addModel(model2, "b_tweede_model.cellml"));
+    EXPECT_TRUE(importer->addModel(model3, "c_troisieme_model.cellml"));
+
+    // Add another under the first key, expect false.
+    EXPECT_FALSE(importer->addModel(model4, "a_model_the_first.cellml"));
+    EXPECT_EQ(size_t(3), importer->libraryCount());
+
+    // Access library by key.
+    EXPECT_EQ(model1, importer->library("a_model_the_first.cellml"));
+    EXPECT_EQ(model2, importer->library("b_tweede_model.cellml"));
+    EXPECT_EQ(model3, importer->library("c_troisieme_model.cellml"));
+    EXPECT_EQ(nullptr, importer->library("d_tuawha.cellml"));
+
+    // Access library by index.
+    EXPECT_EQ(model1, importer->library(0));
+    EXPECT_EQ(model2, importer->library(1));
+    EXPECT_EQ(model3, importer->library(2));
+    EXPECT_EQ(nullptr, importer->library(999));
+}
+
+TEST(Importer, tryingStuffOut)
+{
+    // This is a test to figure out whether we need to restrict the library keys to absolute URLs, or
+    // whether they could be any string.  This test creates a collection of models from different sources,
+    // some which need imports, some not, and uses them to resolve imports on another model.
+    auto parser = libcellml::Parser::create();
+
+    // Parsing concrete units and components.
+    auto concreteUnits = parser->parseModel(fileContents("importer/units_concrete.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto concreteComponents = parser->parseModel(fileContents("importer/components_concrete.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    // Parsing a model which imports components and units from files called units_source.cellml
+    // and components_source.cellml respectively.  These "hidden" files will also be added to the importer
+    // library when they're resolved.
+    auto importedUnits = parser->parseModel(fileContents("importer/units_imported.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto importedComponents = parser->parseModel(fileContents("importer/components_imported.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    // Create models through the API. These don't need import resolution.
+    auto localConcreteUnits = libcellml::Model::create("localConcreteUnits");
+    localConcreteUnits->addUnits(libcellml::Units::create("units5"));
+
+    // Add all the concrete models to the Importer. These don't need resolving as they have no imports of their own.
+    auto importer = libcellml::Importer::create();
+    importer->addModel(concreteUnits, "i_dont_exist_yet.cellml"); // add with key that's not a file (yet).
+    importer->addModel(concreteComponents, "https://www.example.com/myComponents.cellml"); // add with a key that is a remote URL
```

Comment:
Full stop missing.

Reply:
Might also want to capitalise the first word (same for the previous comment).

Reply:
c8650847